saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/RelyingPartyRegistrationResolver.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.provider.service.web;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.security.saml2.provider.service.registration.RelyingPartyRegistration;
+
+/**
+ * A contract for resolving a {@link RelyingPartyRegistration} from the HTTP request
+ *
+ * @author Josh Cummings
+ * @since 5.6
+ */
+public interface RelyingPartyRegistrationResolver extends Converter<HttpServletRequest, RelyingPartyRegistration> {
+
+	@Override
+	default RelyingPartyRegistration convert(HttpServletRequest request) {
```

Comment:
I believe this raises a passivity concern. For example, a current construction of `Saml2MetadataFilter` could look like:

```java
new Saml2MetadataFilter(
    new DefaultRelyingPartyRegistrationResolver(repository), new OpenSamlMetadataResolver())
```

If `RelyingPartyRegistrationResolver` does not extend `Converter`, then `Saml2MetadataFilter` will need another constructor, one that takes `RelyingPartyRegistrationResolver`. Since `DefaultRelyingPartyRegistrationResolver` extends both, the constructor call is ambiguous, causing a compilation error.

One way to address this would be to add a static factory method `Saml2MetadataFilter#construct(RelyingPartyRegistrationResolver, Saml2MetadataResolver)`. This method would be necessary in other classes like `Saml2AuthenticationTokenConverter`.

Or, if we add the new constructors and deprecate the old ones, those who upgrade can change to the other constructor by adding a cast like so:

```java
new Saml2MetadataFilter((RelyingPartyRegistrationResolver) 
    new DefaultRelyingPartyRegistrationResolver(repository), new OpenSamlMetadataResolver())
```

or by splitting up the line (more readable anyway):

```java
RelyingPartyRegistrationResolver relyingPartyRegistrationResolver = 
    new DefaultRelyingPartyRegistrationResolver(repository);
new Saml2MetadataFilter(relyingPartyRegistrationResolver, new OpenSamlMetadataResolver())
```

The initial construction I listed may be uncommon and is certainly less readable -- what are your thoughts?